Bottom: 30e99e2cfb0dbf07e3f25a9e556863894f358734
Top:    1b245bb0d77a5525617d6a902798b750e478e1bc
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-02-02 22:26:22 +0000

Refresh of expose-each-region-separately

---

diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 17ecae2..904bc60 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -951,6 +951,9 @@ gboolean
 ghid_port_drawing_area_expose_event_cb (GtkWidget * widget,
 					GdkEventExpose * ev, GHidPort * port)
 {
+  GdkRectangle *rectangles;
+  int n_rectangles;
+  int i;
   BoxType region;
   int eleft, eright, etop, ebottom;
   extern HID ghid_hid;
@@ -1073,7 +1076,26 @@ ghid_port_drawing_area_expose_event_cb (GtkWidget * widget,
                              -gport->view_x0,
                 ghid_flip_y ? gport->view_y0 - PCB->MaxHeight :
                              -gport->view_y0, 0);
-  hid_expose_callback (&ghid_hid, &region, 0);
+
+  gdk_region_get_rectangles (ev->region, &rectangles, &n_rectangles);
+  for (i = 0; i < n_rectangles; i++) {
+    int x, y, width, height;
+
+    x = rectangles[i].x;
+    y = rectangles[i].y;
+    width = rectangles[i].width;
+    height = rectangles[i].height;
+
+    region.X1 = MIN (Px (x), Px (x + width + 1));
+    region.X2 = MAX (Px (x), Px (x + width + 1));
+    region.Y1 = MIN (Py (y), Py (y + height + 1));
+    region.Y2 = MAX (Py (y), Py (y + height + 1));
+
+    hid_expose_callback (&ghid_hid, &region, 0);
+  }
+  g_free (rectangles);
+
+
   hidgl_flush_triangles (&buffer);
   glPopMatrix ();
